File: backend/src/main/java/com/lsoft/aguapotable/service/LecturaService.java
package com.lsoft.aguapotable.service;

import com.lsoft.aguapotable.domain.entity.Lectura;
import com.lsoft.aguapotable.domain.entity.Medidor;
import com.lsoft.aguapotable.domain.repository.LecturaRepository;
import com.lsoft.aguapotable.domain.repository.MedidorRepository;
import com.lsoft.aguapotable.dto.request.LecturaRequest;
import com.lsoft.aguapotable.exception.BusinessException;
import com.lsoft.aguapotable.exception.ResourceNotFoundException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;

@Service
@RequiredArgsConstructor
@Slf4j
public class LecturaService {

    private final LecturaRepository lecturaRepository;
    private final MedidorRepository medidorRepository;
    private final AlertaService alertaService;

    public List<Lectura> getPorMedidor(Long medidorId) {
        return lecturaRepository.findByMedidorIdOrderByAnioDescMesDesc(medidorId);
    }

    public List<Lectura> getPorPeriodo(Long comunidadId, Integer anio, Integer mes) {
        return lecturaRepository.findByComunidadAndPeriodo(comunidadId, anio, mes);
    }

    @Transactional
    public Lectura registrar(LecturaRequest request, Long comunidadId) {
        Medidor medidor = medidorRepository.findById(request.getMedidorId())
                .orElseThrow(() -> new ResourceNotFoundException("Medidor", request.getMedidorId()));

        // Validar que el medidor pertenece a la comunidad
        if (!medidor.getSuscriptor().getComunidad().getId().equals(comunidadId)) {
            throw new ResourceNotFoundException("Medidor", request.getMedidorId());
        }

        if (request.getLecturaActual() < request.getLecturaAnterior()) {
            throw new BusinessException("La lectura actual no puede ser menor que la anterior");
        }

        LocalDate fecha = request.getFechaLectura() != null ? request.getFechaLectura() : LocalDate.now();
        int anio = fecha.getYear();
        int mes = fecha.getMonthValue();

        // Validar que no exista lectura previa para el mismo período
        if (lecturaRepository.findByMedidorIdAndAnioAndMes(
                request.getMedidorId(), anio, mes).isPresent()) {
            throw new BusinessException("Ya existe una lectura para el medidor en " + anio + "/" + mes);
        }

        double consumo = request.getLecturaActual() - request.getLecturaAnterior();

        Lectura lectura = Lectura.builder()
                .medidor(medidor)
                .anio(anio)
                .mes(mes)
                .lecturaAnterior(request.getLecturaAnterior())
                .lecturaActual(request.getLecturaActual())
                .consumoM3(consumo)
                .fechaLectura(fecha)
                .observaciones(request.getObservaciones())
                .estimada(request.getEstimada() != null && request.getEstimada())
                .build();

        Lectura saved = lecturaRepository.save(lectura);

        // Detectar consumo anómalo
        try {
            alertaService.detectarConsumoAnomalo(saved);
        } catch (Exception e) {
            log.warn("Error al detectar consumo anómalo: {}", e.getMessage());
        }

        return saved;
    }
}

File: backend/src/lectura/lectura.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Lectura, Medidor } from '../common/entities';
import { LecturaRequest } from './dto/lectura.dto';
import {
  BusinessException,
  ResourceNotFoundException,
} from '../common/exceptions';
import { AlertaService } from '../alerta/alerta.service';

@Injectable()
export class LecturaService {
  private readonly logger = new Logger(LecturaService.name);

  constructor(
    @InjectRepository(Lectura)
    private readonly repo: Repository<Lectura>,
    @InjectRepository(Medidor)
    private readonly medidorRepo: Repository<Medidor>,
    private readonly alertaService: AlertaService,
  ) {}

  async getPorMedidor(medidorId: number): Promise<Lectura[]> {
    return this.repo.find({
      where: { medidorId },
      order: { anio: 'DESC', mes: 'DESC' },
    });
  }

  async getPorPeriodo(
    comunidadId: number,
    anio: number,
    mes: number,
  ): Promise<Lectura[]> {
    return this.repo
      .createQueryBuilder('l')
      .innerJoinAndSelect('l.medidor', 'm')
      .innerJoin('m.suscriptor', 's')
      .where('s.comunidad_id = :comunidadId', { comunidadId })
      .andWhere('l.anio = :anio', { anio })
      .andWhere('l.mes = :mes', { mes })
      .getMany();
  }

  async registrar(dto: LecturaRequest, comunidadId: number): Promise<Lectura> {
    const medidor = await this.medidorRepo.findOne({
      where: { id: dto.medidorId },
      relations: ['suscriptor'],
    });
    if (!medidor)
      throw new ResourceNotFoundException('Medidor', dto.medidorId);
    if (medidor.suscriptor.comunidadId !== Number(comunidadId)) {
      throw new ResourceNotFoundException('Medidor', dto.medidorId);
    }

    if (dto.lecturaActual < dto.lecturaAnterior) {
      throw new BusinessException(
        'La lectura actual no puede ser menor que la anterior',
      );
    }

    const fechaStr =
      dto.fechaLectura || new Date().toISOString().split('T')[0];
    const fecha = new Date(fechaStr);
    const anio = fecha.getFullYear();
    const mes = fecha.getMonth() + 1;

    const existe = await this.repo.findOne({
      where: { medidorId: dto.medidorId, anio, mes },
    });
    if (existe) {
      throw new BusinessException(
        `Ya existe una lectura para el medidor en ${anio}/${mes}`,
      );
    }

    const consumo = dto.lecturaActual - dto.lecturaAnterior;

    const lectura = this.repo.create({
      medidorId: dto.medidorId,
      anio,
      mes,
      lecturaAnterior: dto.lecturaAnterior,
      lecturaActual: dto.lecturaActual,
      consumoM3: consumo,
      fechaLectura: fechaStr,
      observaciones: dto.observaciones ?? null,
      estimada: dto.estimada ?? false,
    });

    const saved = await this.repo.save(lectura);

    try {
      await this.alertaService.detectarConsumoAnomalo(saved);
    } catch (e: any) {
      this.logger.warn('Error al detectar consumo anómalo: ' + e.message);
    }

    return saved;
  }
}
